### PR TITLE
linera net up --kubernetes multiple validators

### PR DESCRIPTION
**Pull request by andresilva91**
## Motivation

We need to be able to run multiple validator for feature parity with `linera net up` as well as being able to pass the end to end tests eventually

## Proposal

Implement that functionality

## Test Plan

WIP

### CODE REVIEW

File: linera-service/src/cli_wrappers/local_kubernetes_net.rs [hunk 1]
```diff
@@ -24,8 +24,9 @@ use tokio::process::Command;
 pub struct LocalKubernetesNetConfig {
     pub network: Network,
     pub testing_prng_seed: Option<u64>,
+    pub num_initial_validators: usize,
+    pub num_shards: usize,
     pub binaries_dir: Option<PathBuf>,
-    pub cluster_id: Option<u32>,
```
Comment (andresilva91) [inline linera-service/src/cli_wrappers/local_kubernetes_net.rs:28]: Since we're gonna be creating multiple kind clusters now, passing the `cluster_id` doesn't make sense anymore since we'll have to create multiple of these

File: linera-service/src/cli_wrappers/kind.rs [hunk 2]
```diff
@@ -11,12 +11,9 @@ pub struct KindCluster {
 }
 
 impl KindCluster {
-    pub async fn new(cluster_id: Option<u32>) -> Result<Self> {
-        let cluster = match cluster_id {
-            Some(id) => Self { id },
-            None => Self {
-                id: Self::get_random_cluster_id(),
-            },
+    pub async fn new() -> Result<Self> {
```
Comment (ma2bd) [inline linera-service/src/cli_wrappers/kind.rs:14]: `create`?

File: linera-service/src/cli_wrappers/kubectl.rs [hunk 3]
```diff
@@ -6,23 +6,44 @@ use anyhow::{Context, Result};
...
     pub async fn port_forward(&mut self, pod_name: &str, ports: &str) -> Result<()> {
-        self.port_forward_child = Some(
-            Command::new("kubectl")
-                .arg("port-forward")
-                .arg(pod_name)
-                .arg(ports)
-                .spawn()
-                .context("Port forwarding failed")?,
-        );
+        let port_forward_child = Command::new("kubectl")
```
Comment (christos-h) [inline linera-service/src/cli_wrappers/kubectl.rs:32]: How does this command know which context to use?

File: linera-service/src/cli_wrappers/kubectl.rs [hunk 4]
```diff
@@ -6,23 +6,44 @@ use anyhow::{Context, Result};
...
-        Self { port_forward_child }
+    pub fn new(port_forward_children: Option<Vec<Child>>) -> Self {
+        Self {
+            port_forward_children,
+        }
+    }
+
+    // If there are multiple clusters, make sure you always properly set the context
+    // before doing kubectl commands, otherwise you might be querying the wrong cluster
+    pub async fn config_use_context(&mut self, cluster_id: u32) -> Result<()> {
```
Comment (christos-h) [inline linera-service/src/cli_wrappers/kubectl.rs:21]: This sets the context for the client at the user level right?

So if I connected with my kubectl to cluster A, then I run `linera net up --k8s`, the I resume work with kubectl, it will be connected to a different cluster right?

So we either need a way to set the context just for the command - or find a way to reset it after we are done.

Furthermore, if you are going to be re-using the context it feels like this should be a field of `KubectlInstance`. Something like:


```rust
struct KubectlInstance {
    root_cmd: Command
}

impl KubectlInstance {
  fn new(id: u32) -> Self {
      Self {
         root_cmd: Command::new("kubectl").arg("context").arg(id)
      }
  }
}
```

And then re-use `root_cmd` for subsequent commands like port-forwarding.

File: linera-service/src/cli_wrappers/kubectl.rs (hunk 3 above)
Comment (christos-h) [reply linera-service/src/cli_wrappers/kubectl.rs:32]: See commend above.

File: linera-service/src/cli_wrappers/kind.rs (hunk 2 above)
Comment (christos-h) [inline linera-service/src/cli_wrappers/kind.rs:14]: `KindCluster` should also have a context (see commend for `KubectlWrapper`).

**Review by christos-h — CHANGES_REQUESTED** [review]

File: linera-service/src/cli_wrappers/kubectl.rs (hunk 4 above)
Comment (ma2bd) [reply linera-service/src/cli_wrappers/kubectl.rs:21]: I wouldn't put the command in the struct, though. (See the existing CLI wrappers)

File: kubernetes/linera-validator/values-local.yaml [hunk 5]
```diff
@@ -5,6 +5,7 @@ lineraImage: linera-test:latest
 lineraImagePullPolicy: Never
 logLevel: "debug"
 proxyPort: 19100
+numShards: 10 # This can be set by the CLI `--set numShards=4`
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply kubernetes/linera-validator/values-local.yaml:8]: PR on the way

File: linera-service/src/cli_wrappers/kind.rs [hunk 6]
```diff
@@ -36,6 +33,10 @@ impl KindCluster {
         Ok(())
     }
 
+    pub fn get_id(&self) -> u32 {
```
[the remark this answers is not in the document]
Comment (andresilva91) [reply linera-service/src/cli_wrappers/kind.rs:36]: TIL! Thank you!

File: linera-service/src/cli_wrappers/kind.rs (hunk 2 above)
Comment (andresilva91) [reply linera-service/src/cli_wrappers/kind.rs:14]: Commented on this below

File: linera-service/src/cli_wrappers/kubectl.rs (hunk 3 above)
Comment (andresilva91) [reply linera-service/src/cli_wrappers/kubectl.rs:32]: Commented above

File: linera-service/src/cli_wrappers/kubectl.rs (hunk 4 above)
Comment (christos-h) [reply linera-service/src/cli_wrappers/kubectl.rs:21]: > So this is actually trickier. You can't put the command in the struct, because you can't reuse the command.

That's fair enough - as MB said, you can have a method which returns a new instance of the `Command` every time.

> EDIT: it actually seems like you can set the context per kubectl command. 

Cool so let's do that.

You can [also use the `KUBECONFIG` env variable](https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#set-the-kubeconfig-environment-variable) which helm also respects.

Basically - what I really want to avoid is us misconfiguring peoples' (and our own) `kubectl` client if possible.

If it really isn't - then we need a plan B.

Comment (andresilva91) [reply linera-service/src/cli_wrappers/kubectl.rs:21]: > as MB said, you can have a method which returns a new instance of the Command every time.
Sounds good. That makes sense just for kubectl and helm though, because we're passing context to those. The other ones will have no arguments in common, so we can probably just keep them inlined IMHO

In regards to helm, I just saw here (for some reason missed when I looked earlier) that helm has `--kube-context`, so I can just use that for it, then we should be good :)

**Review by ma2bd — APPROVED** [review]
